### PR TITLE
fixed a bug related to Task Ownership

### PR DESCRIPTION
**Pull request by abdallahdak7**

### CODE REVIEW

File: src/auth/admin.controller.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import { Params } from '../models/params.model';
+import { AdminService } from './admin.service';
+
+@Controller('admin')
+@UseGuards(AuthGuard())
+export class AdminController {
+
+    public constructor(
+        private readonly adminService: AdminService,
+
```
Comment (Khousheish) [inline src/auth/admin.controller.ts]: remove space

File: src/auth/admin.controller.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    ) { }
+
+    @Post('approve/:id')
+    public async approveCustomer(
+        @Param('id', ParseIntPipe) id: number,
+        @GetUser() user: User,
+    ): Promise<User> {
+
```
Comment (Khousheish) [inline src/auth/admin.controller.ts]: remove

File: src/auth/admin.controller.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    @Get('trip-organizers')
+
+    // tslint:disable-next-line: prefer-function-over-method
+    public async getAllTripOrganizers(
+        @GetUser() user: User,
+        @Query() params: Params,
+    ): Promise<User[]> {
+
```
Comment (Khousheish) [inline src/auth/admin.controller.ts]: remove

File: src/auth/admin.controller.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        return AdminService.getAllTripOrganizers(user, params);
+    }
+
```
Comment (Khousheish) [inline src/auth/admin.controller.ts]: remove

File: src/auth/admin.service.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+import { Params } from '../models/params.model';
+import { User } from './user.entity';
+import { UserRepository } from './user.repository';
+
+@Injectable()
+export class AdminService {
+
+    public constructor(private readonly userRepository: UserRepository) {
+
```
Comment (Khousheish) [inline src/auth/admin.service.ts:12]: remove

File: src/auth/admin.service.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    ): Promise<User> {
+        const approvedUser: User = await this.userRepository.approveCustomer(id, user);
+
+        return approvedUser;
+    }
+    public static async getAllTripOrganizers(
+        user: User,
+        params: Params,
+    ): Promise<User[]> {
+
```
Comment (Khousheish) [inline src/auth/admin.service.ts]: remove

File: src/auth/admin.controller.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        const approvedUser: User = await this.adminService.approveCustomer(id, user);
+
+        return approvedUser;
+    }
+
+    @Get('trip-organizers')
+
+    // tslint:disable-next-line: prefer-function-over-method
+    public async getAllTripOrganizers(
+        @GetUser() user: User,
```
Comment (Khousheish) [inline src/auth/admin.controller.ts]: all files are indented with 4 spaces please refactor to use 2

Comment (Khousheish) [reply src/auth/admin.controller.ts]: check all files

**Review by Khousheish — CHANGES_REQUESTED** [review]

File: src/auth/get-user.decorator.ts [hunk 8]
```diff
@@ -1,6 +1,8 @@
 import { createParamDecorator, ExecutionContext } from '@nestjs/common';
+
 import { User } from './user.entity';
 
 export const GetUser = createParamDecorator((data: unknown, executionContext: ExecutionContext): User => {
+
```
Comment (Khousheish) [inline src/auth/get-user.decorator.ts]: remove

File: src/auth/user.repository.ts [hunk 9]
```diff
@@ -86,7 +48,105 @@ export class UserRepository extends Repository<User> {
     return null;
   }
 
-  private async hashPassword(password: string, salt: string): Promise<string> {
+  public async approveCustomer(
+    id: number,
+    user: User,
+  ): Promise<User> {
+
```
Comment (Khousheish) [inline src/auth/user.repository.ts]: remove

File: src/auth/user.repository.ts [hunk 10]
```diff
@@ -86,7 +48,105 @@ export class UserRepository extends Repository<User> {
...
+      const approvedUser: User = await getManager()
+        .createQueryBuilder(User, 'user')
+        .where({ id })
+        .getOne();
+
+      return approvedUser;
+    }
+    throw new UnauthorizedException('Only Admins can approve a user.');
+
```
Comment (Khousheish) [inline src/auth/user.repository.ts]: remove

File: src/auth/user.repository.ts [hunk 11]
```diff
@@ -86,7 +48,105 @@ export class UserRepository extends Repository<User> {
...
+  }
+
+  public static async getAllTripOrganizers(
+    user: User,
+    params: Params,
+  ): Promise<User[]> {
+
```
Comment (Khousheish) [inline src/auth/user.repository.ts]: remove

File: src/auth/user.repository.ts [hunk 12]
```diff
@@ -86,7 +48,105 @@ export class UserRepository extends Repository<User> {
...
+      if (error.constraint === 'UQ_e12875dfb3b1d92d7d7c5377e22') {
+        throw new ConflictException('Email already exists.');
+      }
+      if (error.constraint === 'UQ_78a916df40e02a9deb1c4b75edb') {
+        throw new ConflictException('Username already exists.');
+      }
+      if (error.constraint === 'UQ_c1756d987198666d8b02af03439') {
+        throw new ConflictException('Phone Number already exists.');
+      }
+      // tslint:disable-next-line: no-console
```
Comment (Khousheish) [inline src/auth/user.repository.ts]: remove these 2 lines

File: src/auth/user.repository.ts [hunk 13]
```diff
@@ -86,7 +48,105 @@ export class UserRepository extends Repository<User> {
...
+    delete user.password;
+    delete user.salt;
+
+    return user;
+  }
+
+  private static async hashPassword(password: string, salt: string): Promise<string> {
     return bcrypt.hash(password, salt);
   }
+
```
Comment (Khousheish) [inline src/auth/user.repository.ts]: remove

File: src/trips/customer.trips.controller.ts [hunk 14]
```diff
@@ -0,0 +1,23 @@
+
```
Comment (Khousheish) [inline src/trips/customer.trips.controller.ts]: remove

File: src/trips/dto/create-trip.dto.ts [hunk 15]
```diff
@@ -1,9 +1,18 @@
...
+  @IsString()
+  public description: string;
+
+  @IsOptional()
+  @IsString()
+  public image: string;
+
+  @IsOptional()
+  public date: Date;
+
```
Comment (Khousheish) [inline src/trips/dto/create-trip.dto.ts]: remove

File: src/trips/trip.entity.ts [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+
+  @Column({ nullable: true })
+  public image: string;
+
+  @Column({ nullable: true })
+  public description: string;
+
+  @Column({ nullable: true })
+  public date: Date;
+
```
Comment (Khousheish) [inline src/trips/trip.entity.ts]: remove

File: src/trips/trip.entity.ts [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+  public name: string;
+
+  // tslint:disable-next-line: typedef
+  @ManyToOne(type => User, user => user.trips, { eager: false })
+  public user: User;
+
+  @Column()
+  public userId: number;
+
+  // tslint:disable-next-line: typedef
```
Comment (Khousheish) [inline src/trips/trip.entity.ts:21]: cant you add the type def to these? they seem simple

File: src/trips/trip.repository.ts [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+import { UserType } from '../enums/user-type.enum';
+import { Params } from '../models/params.model';
+import { CreateTripDto } from './dto/create-trip.dto';
+import { Trip } from './trip.entity';
+
+@EntityRepository(Trip)
+export class TripRepository extends Repository<Trip>{
+
+  public async customerGetAllTrips(user: User): Promise<Trip[]> {
+
```
Comment (Khousheish) [inline src/trips/trip.repository.ts]: remove

File: src/trips/trip.repository.ts [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+    return trips;
+  }
+
+  public async getAllTrips(
+    user: User,
+    params: Params,
+  ): Promise<Trip[]> {
+    let query: SelectQueryBuilder<Trip>;
+    if (user.userType === UserType.tripOrganizer) {
```
Comment (Khousheish) [inline src/trips/trip.repository.ts:27]: space

File: src/trips/trips.controller.spec.ts [hunk 20]
```diff
@@ -1,18 +1,21 @@
 import { Test, TestingModule } from '@nestjs/testing';
 
-import { TripsController } from './trips.controller';
+import { CustomerTripsController } from './customer.trips.controller';
 
+// tslint:disable-next-line: typedef
```
Comment (Khousheish) [inline src/trips/trips.controller.spec.ts]: type void add and remove disable

File: src/trips/trips.controller.spec.ts [hunk 21]
```diff
@@ -1,18 +1,21 @@
...
 describe('Trips Controller', () => {
-  let controller: TripsController;
+  let controller: CustomerTripsController;
 
-  beforeEach(async () => {
+  // tslint:disable-next-line: typedef
```
Comment (Khousheish) [inline src/trips/trips.controller.spec.ts]: check rest of files

File: src/trips/trips.controller.ts [hunk 22]
```diff
@@ -1,49 +1,56 @@
...
   ) { }
 
   @Get()
-  public async getAllTrips(@GetUser() tuser: User): Promise<Trip[]> {
-    return this.tripsService.getAllTrips(tuser);
+  public async getAllTrips(
+    @GetUser() user: User,
+    @Query() params: Params,
+  ): Promise<Trip[]> {
+
```
Comment (Khousheish) [inline src/trips/trips.controller.ts]: remove

File: src/trips/trips.controller.ts [hunk 23]
```diff
@@ -1,49 +1,56 @@
...
 
   @Get('/:id')
   public async getTripById(
     @Param('id', ParseIntPipe) id: number,
-    @GetUser() tuser: User,
-    ): Promise<Trip> {
-    return this.tripsService.getTripById(id, tuser);
+    @GetUser() user: User,
+  ): Promise<Trip> {
+
```
Comment (Khousheish) [inline src/trips/trips.controller.ts]: remove

File: src/trips/trips.controller.ts [hunk 24]
```diff
@@ -1,49 +1,56 @@
...
   @Post()
   @UsePipes(ValidationPipe)
+  // tslint:disable-next-line: prefer-function-over-method
   public async createTrip(
     @Body() createTripDto: CreateTripDto,
-    @GetUser() tuser: User,
+    @GetUser() user: User,
   ): Promise<Trip> {
-    return this.tripsService.createTrip(createTripDto, tuser);
+
```
Comment (Khousheish) [inline src/trips/trips.controller.ts]: remove

File: src/trips/trips.controller.ts [hunk 25]
```diff
@@ -1,49 +1,56 @@
...
 
   @Delete('/:id')
   public async deleteTrip(
     @Param('id', ParseIntPipe) id: number,
-    @GetUser() tuser: User,
-    ): Promise<void> {
-    return this.tripsService.deleteTrip(id, tuser);
+    @GetUser() user: User,
+  ): Promise<Trip> {
+
```
Comment (Khousheish) [inline src/trips/trips.controller.ts]: remove

File: src/trips/trips.service.spec.ts [hunk 26]
```diff
@@ -1,17 +1,20 @@
 import { Test, TestingModule } from '@nestjs/testing';
 import { TripsService } from './trips.service';
 
+// tslint:disable-next-line: typedef
 describe('TripsService', () => {
   let service: TripsService;
 
-  beforeEach(async () => {
+  // tslint:disable-next-line: typedef
```
Comment (Khousheish) [inline src/trips/trips.service.spec.ts]: fix in all the file

File: src/trips/trips.service.ts [hunk 27]
```diff
@@ -16,39 +16,61 @@ export class TripsService {
     private readonly tripRepository: TripRepository,
   ) { }
 
-  public async getAllTrips(user: User): Promise<Trip[]> {
+  public async customerGetAllTrips(user: User): Promise<Trip[]> {
+
```
Comment (Khousheish) [inline src/trips/trips.service.ts]: remove

File: src/trips/trips.service.ts [hunk 28]
```diff
@@ -16,39 +16,61 @@ export class TripsService {
...
+    return this.tripRepository.customerGetAllTrips(user);
+  }
+
+  public async getAllTrips(
+    user: User,
+    params: Params,
+  ): Promise<Trip[]> {
```
Comment (Khousheish) [inline src/trips/trips.service.ts]: remove

File: src/trips/trips.service.ts [hunk 29]
```diff
@@ -16,39 +16,61 @@ export class TripsService {
...
-    return this.tripRepository.createTrip(createTripDto, user);
-  }
-
   public async deleteTrip(
     id: number,
     user: User,
-  ): Promise<void> {
-    const result: DeleteResult = await this.tripRepository.delete({ id, userId: user.id });
+  ): Promise<Trip> {
+
```
Comment (Khousheish) [inline src/trips/trips.service.ts]: remove

**Review by Khousheish — CHANGES_REQUESTED** [review]

**Review by KbMstafa — APPROVED** [review]

**Review by Khousheish — APPROVED** [review]

**Pull request merged**